packages/fast-tooling-react/src/css-property-editor/property-editor.tsx

```diff
@@ -1,237 +1,365 @@
+                ] = this.editData[key];
+            } else {
+                newData[key] = this.editData[key];
+            }
+        });
+
+        this.editData = newData;
         this.setState({
-            key: e.target.value,
+            currentEditRowUncommittedCSSPropertyKey: null,
         });
+
+        this.handleCSSUpdate(newData);
     };
 
-    private handleNewPropertyValueChange = (
-        e: React.ChangeEvent<HTMLInputElement>
-    ): void => {
+    /**
+     * Row gained focus
+     */
+    private handleRowFocus = (rowKey: string, rowIndex: number): void => {
+        if (this.currentEditRowIndex !== rowIndex) {
+            this.currentEditRowReactKey = null;
+            this.currentEditRowIndex = rowIndex;
+        }
+    };
+
+    /**
+     * Key input has lost focus
+     */
+    private handleCommitKeyEdit = (rowKey: string, rowIndex: number): void => {
+        if (this.state.currentEditRowUncommittedCSSPropertyKey !== null) {
+            if (this.state.currentEditRowUncommittedCSSPropertyKey === "") {
+                this.deleteRow(rowIndex);
+            } else {
+                this.commitCSSKeyEdit();
+            }
+        } else if (rowKey === "") {
+            this.deleteRow(rowIndex);
+        }
+    };
+
+    /**
+     * Row has lost focus
+     */
+    private handleRowBlur = (rowKey: string, rowIndex: number): void => {
+        if (this.currentEditRowIndex !== -1 && this.editData[rowKey] === "") {
+            this.deleteRow(rowIndex);
+        }
+        this.currentEditRowReactKey = null;
         this.setState({
-            value: e.target.value,
+            currentEditRowUncommittedCSSPropertyKey: null,
         });
     };
 
-    private handleNewPropertyValueKeyDown = (
-        e: React.KeyboardEvent<HTMLInputElement>
-    ): void => {
-        if (e.keyCode === KeyCodes.tab) {
-            if (this.state.key !== "" && this.state.value !== "") {
-                e.preventDefault();
+    /**
+     * Clicks on a row outside the inputs add a row following that row
+     */
+    private handleClickToInsert = (rowKey: string, rowIndex: number): void => {
+        this.createRow(rowIndex + 1);
+    };
+
+    /**
+     * Add a row at the end of the list when there are clicks outside of an existing row
+     */
+    private handleClick = (e: React.MouseEvent<HTMLDivElement>): void => {
+        if (e.defaultPrevented) {
+            return;
+        }
+        this.createRow(Object.keys(this.editData).length);
+    };
+
+    /**
+     * Component got focus without any data rows, so add an empty one
+     */
+    private handleEmptyFocus = (): void => {
+        this.createRow(0);
+    };
+
+    /**
+     * Create a new row at the insertion index
+     */
+    private createRow = (insertionIndex: number): void => {
+        const newData: CSSProperties = {};
+        const keys: string[] = Object.keys(this.editData);
+
+        for (let i: number = 0; i <= keys.length; i = i + 1) {
```

Comment:
Instead of `i = i + 1` we prefer `i++`

You should also store the value of length otherwise it will hit length each time which is a slight perf hit. See https://jsperf.com/array-length-in-loop/25

You can do this by:

`for (let i: number = 0, keysLength: number = keys.length; i <= keysLength; i++)`

Reply:
yep,  didn't know you could declare multiple things in that first part of the statement